ignite/metrics/precision.py

```diff
@@ -10,33 +11,116 @@
+    - For multilabel cases, `y` and `y_pred` must have same shape of (batch_size, num_categories, ...).
+    For binary or multiclass cases, if `average` is True, returns the unweighted average across all classes.
     Otherwise, returns a tensor with the precision for each class.
+    For multilabel cases `average` is True and returns the unweighted average across all samples.
     """
-    def __init__(self, average=False, output_transform=lambda x: x):
-        super(Precision, self).__init__(output_transform)
+    def __init__(self, output_transform=lambda x: x, is_multilabel=False, average=False, threshold_function=None):
         self._average = average
+        if is_multilabel:
+            if threshold_function is None:
+                self._threshold = torch.round
+            else:
+                if callable(threshold_function):
+                    self._threshold = threshold_function
+                else:
+                    raise ValueError("threshold_function must be a callable function.")
+            if not self._average:
+                warnings.warn('average should be True for multilabel cases. Precision._average updated'
+                              ' to True. Average is calculated across samples, instead of classes.', UserWarning)
+                self._average = True
+            self.update = self._update_multilabel
+        else:
+            self.update = self._update_multiclass
+        super(Precision, self).__init__(output_transform=output_transform)
 
     def reset(self):
         self._all_positives = None
         self._true_positives = None
 
     def update(self, output):
+        pass
+
+    def compute(self):
+        if self._all_positives is None:
+            raise NotComputableError('Precision must have at least one example before it can be computed')
+
+        result = self._true_positives / self._all_positives
+        result[result != result] = 0.0
+        if self._average:
+            return result.mean().item()
+        else:
+            return result
+
+    def _update_multilabel(self, output):
         y_pred, y = output
         dtype = y_pred.type()
+        axis = 1
+
+        if not (y.shape == y_pred.shape and y.ndimension() > 1 and y.shape[1] != 1):
+            raise ValueError("y and y_pred must have same shape of (batch_size, num_categories, ...).")
+
+        if y_pred.ndimension() == 3:
+            # Converts y and y_pred to (-1, num_classes) from N x C x L
+            y_pred = y_pred.transpose(2, 1).contiguous().view(-1, y_pred.size(1))
+            y = y.transpose(2, 1).contiguous().view(-1, y_pred.size(1))
+
+        elif y_pred.ndimension() == 4:
+            # Converts y and y_pred to (-1, num_classes) from N x C x H x W
+            y_pred = y_pred.permute(0, 2, 3, 1).contiguous().view(-1, y_pred.size(1))
+            y = y.permute(0, 2, 3, 1).contiguous().view(-1, y_pred.size(1))
+
+        else:
```

Comment:
Can we remove this too ?

Reply:
Incorporated. Generalized as suggested below.